File: modules/smalltalk/greetings.pl
% prolog

% 
% answer(greeting, en, anonymous) :-
%     rdf(ai:curin, ai:currentTime, TS),
%     before_noon(TS),
%     action(attention, on),
%     say_eoa(en, "Good Morning!").
% answer(greeting, de, anonymous) :-
%     rdf(ai:curin, ai:currentTime, TS),
%     before_noon(TS),
%     action(attention, on),
%     say_eoa(de, "Guten Morgen!").
% 
% answer(greeting, en, anonymous) :-
%     rdf(ai:curin, ai:currentTime, TS),
%     before_evening(TS),
%     after_noon(TS),
%     action(attention, on),
%     say_eoa(en, "Good Day to you!"),
%     action(attention, on),
%     say_eoa(en, "Hello!").
% answer(greeting, de, anonymous) :-
%     rdf(ai:curin, ai:currentTime, TS),
%     before_evening(TS),
%     after_noon(TS),
%     action(attention, on),
%     say_eoa(de, "Guten Tag!"),
%     action(attention, on),
%     say_eoa(de, "Hallo!").
% 
% answer(greeting, en, anonymous) :-
%     rdf(ai:curin, ai:currentTime, TS),
%     after_evening(TS),
%     action(attention, on),
%     say_eoa(en, "Good Evening!").
% answer(greeting, de, anonymous) :-
%     rdf(ai:curin, ai:currentTime, TS),
%     after_evening(TS),
%     action(attention, on),
%     say_eoa(de, "Guten Abend!").
% 
% answer(goodbye, en, anonymous) :-
%     action(attention,off),
%     say_eoa(en, "Goodbye!").
% answer(goodbye, de, anonymous) :-
%     action(attention,off),
%     say_eoa(de, "Auf Wiedersehen!").

%
% hello
%

nlp_greetings_s (en, hello, S) :- hears (en, S, 'greetings').
nlp_greetings_s (en, hello, S) :- hears (en, S, 'good morning').
nlp_greetings_s (en, hello, S) :- hears (en, S, 'hello').
nlp_greetings_s (en, hello, S) :- hears (en, S, 'hallo').
nlp_greetings_s (en, hello, S) :- hears (en, S, 'hi').
nlp_greetings_s (en, hello, S) :- hears (en, S, 'good day').
nlp_greetings_s (en, hello, S) :- hears (en, S, 'morning').
nlp_greetings_s (en, hello, S) :- hears (en, S, 'good evening').
nlp_greetings_s (en, hello, S) :- hears (en, S, 'good night').
nlp_greetings_s (en, hello, S) :- hears (en, S, 'Cooee').
nlp_greetings_s (en, hello, S) :- hears (en, S, 'Cooey').
nlp_greetings_s (en, hello, S) :- hears (en, S, 'hi there').

nlp_greetings_s (de, hello, S) :- hears (de, S, 'grüß dich').
nlp_greetings_s (de, hello, S) :- hears (de, S, 'hallo').
nlp_greetings_s (de, hello, S) :- hears (de, S, 'hi').
nlp_greetings_s (de, hello, S) :- hears (de, S, 'guten morgen').
nlp_greetings_s (de, hello, S) :- hears (de, S, 'guten tag').
nlp_greetings_s (de, hello, S) :- hears (de, S, 'guten abend').
nlp_greetings_s (de, hello, S) :- hears (de, S, 'guten nachmittag').
nlp_greetings_s (de, hello, S) :- hears (de, S, 'gute nacht').
nlp_greetings_s (de, hello, S) :- hears (de, S, 'schönen guten morgen').
nlp_greetings_s (de, hello, S) :- hears (de, S, 'schönen guten tag').
nlp_greetings_s (de, hello, S) :- hears (de, S, 'schönen guten abend').
nlp_greetings_s (de, hello, S) :- hears (de, S, 'schönen guten nachmittag').
nlp_greetings_s (de, hello, S) :- hears (de, S, 'gute nacht').
nlp_greetings_s (de, hello, S) :- hears (de, S, 'tag').
nlp_greetings_s (de, hello, S) :- hears (de, S, 'morgen').
nlp_greetings_s (de, hello, S) :- hears (de, S, 'huhu').

nlp_greetings_r (en, hello, R) :- acts(R, attention(on)), says (en, R, "Hello!").
nlp_greetings_r (en, hello, R) :- acts(R, attention(on)), says (en, R, "Hi!").
nlp_greetings_r (en, hello, R) :- acts(R, attention(on)), says (en, R, "Greetings!").
nlp_greetings_r (en, hello, R) :- acts(R, attention(on)), says (en, R, "Hey!").

nlp_greetings_r (de, hello, R) :- acts(R, attention(on)), says (de, R, "Hallo!").
nlp_greetings_r (de, hello, R) :- acts(R, attention(on)), says (de, R, "Hi!").
nlp_greetings_r (de, hello, R) :- acts(R, attention(on)), says (de, R, "Grüß Dich!").
nlp_greetings_r (de, hello, R) :- acts(R, attention(on)), says (de, R, "Hey!").

nlp_train('smalltalk', en, [[], S1, [], R1]) :-
    self_address(en, S1, _),
    nlp_greetings_s (en, hello, S1),
    nlp_greetings_r (en, hello, R1).
nlp_train('smalltalk', de, [[], S1, [], R1]) :-
    self_address(de, S1, _),
    nlp_greetings_s (de, hello, S1),
    nlp_greetings_r (de, hello, R1).

nlp_test('smalltalk', en, 'hello1', [],
         ['Hi!', 'hello!', [attention(on)]]).
nlp_test('smalltalk', de, 'hello2', [],
         ['Hi!', 'hallo!', [attention(on)]]).
nlp_test('smalltalk', en, 'hello3', [],
         ['Computer, Hi!', 'hello!', [attention(on)]]).
nlp_test('smalltalk', de, 'hello4', [],
         ['Computer, Hi!', 'hallo!', [attention(on)]]).

%
% bye
%

nlp_greetings_s (en, goodbye, S) :- hears (en, S, 'goodbye').
nlp_greetings_s (en, goodbye, S) :- hears (en, S, 'bye').
nlp_greetings_s (en, goodbye, S) :- hears (en, S, 'ciao').
nlp_greetings_s (en, goodbye, S) :- hears (en, S, 'so long').
nlp_greetings_s (en, goodbye, S) :- hears (en, S, 'bye for now').
nlp_greetings_s (en, goodbye, S) :- hears (en, S, 'see ya').
nlp_greetings_s (en, goodbye, S) :- hears (en, S, 'see you later').
nlp_greetings_s (en, goodbye, S) :- hears (en, S, 'till next time').

nlp_greetings_s (de, goodbye, S) :- hears (de, S, 'auf wiedersehen').
nlp_greetings_s (de, goodbye, S) :- hears (de, S, 'tschüss').
nlp_greetings_s (de, goodbye, S) :- hears (de, S, 'ciao').
nlp_greetings_s (de, goodbye, S) :- hears (de, S, 'ade').
nlp_greetings_s (de, goodbye, S) :- hears (de, S, 'bye').
nlp_greetings_s (de, goodbye, S) :- hears (de, S, 'cu').
nlp_greetings_s (de, goodbye, S) :- hears (de, S, 'bis bald').
nlp_greetings_s (de, goodbye, S) :- hears (de, S, 'bis zum nächsten mal').

nlp_greetings_r (en, goodbye, R) :- acts(R, attention(off)), says (en, R, "Bye").
nlp_greetings_r (en, goodbye, R) :- acts(R, attention(off)), says (en, R, "So long").
nlp_greetings_r (en, goodbye, R) :- acts(R, attention(off)), says (en, R, "See you later").
nlp_greetings_r (en, goodbye, R) :- acts(R, attention(off)), says (en, R, "Bye for now").

nlp_greetings_r (de, goodbye, R) :- acts(R, attention(off)), says (de, R, "Ade").
nlp_greetings_r (de, goodbye, R) :- acts(R, attention(off)), says (de, R, "Tschüss").
nlp_greetings_r (de, goodbye, R) :- acts(R, attention(off)), says (de, R, "Bis bald").
nlp_greetings_r (de, goodbye, R) :- acts(R, attention(off)), says (de, R, "Ciao").

nlp_train('smalltalk', en, [[], S1, [], R1]) :-
    nlp_greetings_s (en, goodbye, S1),
    self_address(en, S1, _),
    nlp_greetings_r (en, goodbye, R1).
nlp_train('smalltalk', de, [[], S1, [], R1]) :-
    nlp_greetings_s (de, goodbye, S1),
    self_address(de, S1, _),
    nlp_greetings_r (de, goodbye, R1).

nlp_test('smalltalk', en, 'goodbye1', [],
         ['bye!', 'bye', [attention(off)]]).
nlp_test('smalltalk', de, 'goodbye2', [],
         ['Tschüss computer!', 'Tschüss!', [attention(off)]]).

%
% howdy
%

nlp_greetings_s (en, howdy, S) :- hears (en, S, [["how are you","howdy","how do you do","how are you feeling"], ["today",""], "?"]).
nlp_greetings_s (de, howdy, S) :- hears (de, S, [["wie geht es dir","wie gehts","was geht","wie fühlst du dich"], ["heute",""], "?"]).

nlp_greetings_r (en, howdy, R) :- says (en, R, "Great, thanks. How do you feel today?").
nlp_greetings_r (en, howdy, R) :- says (en, R, "Very well - and you?").
nlp_greetings_r (en, howdy, R) :- says (en, R, "I am doing great, how are you doing?").
nlp_greetings_r (en, howdy, R) :- says (en, R, "Great as always!").
nlp_greetings_r (en, howdy, R) :- says (en, R, "Thanks for asking, I am doing fine. How about you?").

nlp_greetings_r (de, howdy, R) :- says (de, R, "Sehr gut, danke. Und selber?").
nlp_greetings_r (de, howdy, R) :- says (de, R, "Gut, danke. Wie geht es Dir?").
nlp_greetings_r (de, howdy, R) :- says (de, R, "Mir geht es prima, und Dir?").
nlp_greetings_r (de, howdy, R) :- says (de, R, "Mir geht es gut, und selber?").
nlp_greetings_r (de, howdy, R) :- says (de, R, "Super, wie immer!").
nlp_greetings_r (de, howdy, R) :- says (de, R, "Gut, danke der Nachfrage. Wie geht es Dir?").

nlp_train('smalltalk', en, [[], S1, [], R1]) :-
    self_address(en, S1, _),
    nlp_greetings_s (en, howdy, S1),
    nlp_greetings_r (en, howdy, R1).
nlp_train('smalltalk', de, [[], S1, [], R1]) :-
    self_address(de, S1, _),
    nlp_greetings_s (de, howdy, S1),
    nlp_greetings_r (de, howdy, R1).

nlp_test('smalltalk', en, 'howdy1', [],
         ['Computer, how are you?', 'very well and you?', []]).
nlp_test('smalltalk', de, 'howdy2', [],
         ['Computer, wie geht es Dir?', 'Super, wie immer!', []]).

File: modules/smalltalk/greetings.py
#!/usr/bin/env python
# -*- coding: utf-8 -*- 

#
# Copyright 2016, 2017 Guenter Bartsch
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

from base.utils        import hears, says, nlp_base_self_address_s, rdf, entity_label, is_male, is_female
from base.conversation import nlp_base_sayagain_s, nlp_base_isaid_r, nlp_base_whatwasourtopic_s
from base.ner          import ner_learn

#
# hello
#

def nlp_greetings_hello_s(lang, s):

    if lang == 'en':

        res = []

        res.extend(hears ('en', s, ['greetings']))
        res.extend(hears ('en', s, ['good morning']))
        res.extend(hears ('en', s, ['hello']))
        res.extend(hears ('en', s, ['hallo']))
        res.extend(hears ('en', s, ['hi']))
        res.extend(hears ('en', s, ['good day']))
        res.extend(hears ('en', s, ['morning']))
        res.extend(hears ('en', s, ['good evening']))
        res.extend(hears ('en', s, ['good night']))
        res.extend(hears ('en', s, ['Cooee']))
        res.extend(hears ('en', s, ['Cooey']))
        res.extend(hears ('en', s, ['hi there']))

    elif lang == 'de':

        res = []

        res.extend(hears ('de', s, ['grüß dich']))
        res.extend(hears ('de', s, ['hallo']))
        res.extend(hears ('de', s, ['hi']))
        res.extend(hears ('de', s, ['guten morgen']))
        res.extend(hears ('de', s, ['guten tag']))
        res.extend(hears ('de', s, ['guten abend']))
        res.extend(hears ('de', s, ['guten nachmittag']))
        res.extend(hears ('de', s, ['gute nacht']))
        res.extend(hears ('de', s, ['schönen guten morgen']))
        res.extend(hears ('de', s, ['schönen guten tag']))
        res.extend(hears ('de', s, ['schönen guten abend']))
        res.extend(hears ('de', s, ['schönen guten nachmittag']))
        res.extend(hears ('de', s, ['gute nacht']))
        res.extend(hears ('de', s, ['tag']))
        res.extend(hears ('de', s, ['morgen']))
        res.extend(hears ('de', s, ['huhu']))

    else:
        raise Exception ('unsupported language: %s' % lang)

    return res

def nlp_greetings_hello_r(lang, r):

    if lang == 'en':

        return [ 
                 says ('en', r, u'Hello!'    , actions = [['attention', 'on']]),
                 says ('en', r, u'Hi!'       , actions = [['attention', 'on']]),
                 says ('en', r, u'Greetings!', actions = [['attention', 'on']]),
                 says ('en', r, u'Hey!'      , actions = [['attention', 'on']]),
               ] 

    elif lang == 'de':

        return [ 
                 says ('de', r, u'Hallo!'    , actions = [['attention', 'on']]),
                 says ('de', r, u'Hi!'       , actions = [['attention', 'on']]),
                 says ('de', r, u'Grüß Dich!', actions = [['attention', 'on']]),
                 says ('de', r, u'Hey!'      , actions = [['attention', 'on']]),
               ]

    else:
        raise Exception ('unsupported language: %s' % lang)
    
def nlp_greetings_hello(kernal, res):

    # single-round training

    for lang in ['en', 'de']:
        for s1_1 in nlp_base_self_address_s(kernal, lang, []):
            for s1_2 in nlp_greetings_hello_s(lang, s1_1):
                for r1 in nlp_greetings_hello_r(lang, []):
                    res.append((lang, [[], s1_2, [], r1]))

    return res

#
# bye
#

def nlp_greetings_goodbye_s(lang, s):

    if lang == 'en':

        res = []

        res.extend(hears ('en', s, ['goodbye']))
        res.extend(hears ('en', s, ['bye']))
        res.extend(hears ('en', s, ['ciao']))
        res.extend(hears ('en', s, ['so long']))
        res.extend(hears ('en', s, ['bye for now']))
        res.extend(hears ('en', s, ['see ya']))
        res.extend(hears ('en', s, ['see you later']))
        res.extend(hears ('en', s, ['till next time']))

    elif lang == 'de':

        res = []

        res.extend(hears ('de', s, ['auf wiedersehen']))
        res.extend(hears ('de', s, ['tschüss']))
        res.extend(hears ('de', s, ['ciao']))
        res.extend(hears ('de', s, ['ade']))
        res.extend(hears ('de', s, ['bye']))
        res.extend(hears ('de', s, ['cu']))
        res.extend(hears ('de', s, ['bis bald']))
        res.extend(hears ('de', s, ['bis zum nächsten mal']))

    else:
        raise Exception ('unsupported language: %s' % lang)

    return res

def nlp_greetings_goodbye_r(lang, r):

    if lang == 'en':

        return [ 
                 says ('en', r, u'Bye'             , actions = [['attention', 'off']]),
                 says ('en', r, u'So long'         , actions = [['attention', 'off']]),
                 says ('en', r, u'See you later'   , actions = [['attention', 'off']]),
                 says ('en', r, u'Bye for now'     , actions = [['attention', 'off']]),
               ] 

    elif lang == 'de':

        return [ 
                 says ('de', r, u'Ade'             , actions = [['attention', 'off']]),
                 says ('de', r, u'Tschüss'         , actions = [['attention', 'off']]),
                 says ('de', r, u'Bis bald'        , actions = [['attention', 'off']]),
                 says ('de', r, u'Ciao'            , actions = [['attention', 'off']]),
               ]

    else:
        raise Exception ('unsupported language: %s' % lang)
    
def nlp_greetings_goodbye(kernal, res):

    # single-round training

    for lang in ['en', 'de']:
        for s1_1 in nlp_greetings_goodbye_s(lang, []):
            for s1_2 in nlp_base_self_address_s(kernal, lang, s1_1):
                for r1 in nlp_greetings_goodbye_r(lang, []):
                    res.append((lang, [[], s1_2, [], r1]))

    return res


#
# howdy
#
 
def nlp_greetings_howdy_s(lang, s):

    if lang == 'en':

        res = []

        res.extend(hears ('en', s, [["how are you","howdy","how do you do","how are you feeling"], ["today",""], "?"]))

    elif lang == 'de':

        res = []

        res.extend(hears ('de', s, [["wie geht es dir","wie gehts","was geht","wie fühlst du dich"], ["heute",""], "?"]))

    else:
        raise Exception ('unsupported language: %s' % lang)

    return res

def nlp_greetings_howdy_r(lang, r):

    if lang == 'en':

        return [ 
                 says ('en', r, u'Great, thanks. How do you feel today?'),
                 says ('en', r, u'Very well - and you?'),
                 says ('en', r, u'I am doing great, how are you doing?'),
                 says ('en', r, u'Great as always!'),
                 says ('en', r, u'Thanks for asking, I am doing fine. How about you?'),
               ] 

    elif lang == 'de':

        return [ 
                 says ('de', r, u'Sehr gut, danke. Und selber?'),
                 says ('de', r, u'Gut, danke. Wie geht es Dir?'),
                 says ('de', r, u'Mir geht es prima, und Dir?'),
                 says ('de', r, u'Mir geht es gut, und selber?'),
                 says ('de', r, u'Super, wie immer!'),
                 says ('de', r, u'Gut, danke der Nachfrage. Wie geht es Dir?'),
               ]

    else:
        raise Exception ('unsupported language: %s' % lang)

def nlp_greetings_howdy(kernal, res):

    # single-round training

    for lang in ['en', 'de']:
        for s1_1 in nlp_base_self_address_s(kernal, lang, []):
            for s1_2 in nlp_greetings_howdy_s(lang, s1_1):
                for r1 in nlp_greetings_howdy_r(lang, []):
                    res.append((lang, [[], s1_2, [], r1]))

    return res

def nlp_smalltalk_greetings_train(kernal, res):

    res = []

    nlp_greetings_hello(kernal, res)
    nlp_greetings_goodbye(kernal, res)
    nlp_greetings_howdy(kernal, res)

    return res

def nlp_smalltalk_greetings_test(kernal, res):

    res.extend([ 
                ('en', 'hello1',   [], ["Hi!",                        "hello!",             [['attention', 'on']]]),
                ('de', 'hello2',   [], ["Hallo!",                     "hallo!",             [['attention', 'on']]]),
                ('en', 'goodbye1', [], ["Bye!",                       "Bye!",               [['attention', 'off']]]),
                ('de', 'goodbye2', [], ["Tschüss Computer!",          "Tschüss!",           [['attention', 'off']]]),
                ('en', 'howdy1',   [], ["Computer, how are you?",     "very well and you?", []]),
                ('de', 'howdy2',   [], ["Computer, wie geht es Dir?", "Super, wie immer!",  []]),
               ])
